### PR TITLE
Optimize cast(JSON as ROW)

### PR DESCRIPTION
**Pull request by mbasmanova**
Summary:
boost::algorithm::to_lower is very high on the profile. In most cases field names are all-ASCII, hence, we can use more efficient folly::toLowerAscii.

With this optimization a query that's heavy on this cast runs 3x faster.

Differential Revision: D56014538

### CODE REVIEW

File: velox/functions/prestosql/types/JsonType.cpp [hunk 1]
```diff
@@ -856,30 +885,55 @@ struct CastFromJsonTypedImpl {
...
           if (!field.value().is_null()) {
             SIMDJSON_ASSIGN_OR_RAISE(key, field.unescaped_key(true));
-            boost::algorithm::to_lower(key);
-            lowerCaseKeys[key] = field.value();
+            std::string lowerCaseKey = key;
+
+            // boost::algorithm::to_lower is very slow. Use much faster
+            // folly::toLowerAscii if possible.
+            if (allFieldsAreAscii) {
+              folly::toLowerAscii(lowerCaseKey);
```
Comment (Yuhta) [inline velox/functions/prestosql/types/JsonType.cpp:912]: Is `toLowerAscii` safe to be called on non-ascii input?  If not we may need to check if `lowerCaseKey` is ascii as well

Comment (mbasmanova) [reply velox/functions/prestosql/types/JsonType.cpp:912]: I think it is: "Leaves all other characters unchanged, including those with the 0x80 bit set."

```
/**
 * Convert ascii to lowercase, in-place.
 *
 * Leaves all other characters unchanged, including those with the 0x80
 * bit set.
 * @param str String to convert
 * @param length Length of str, in bytes
 */
void toLowerAscii(char* str, size_t length);

inline void toLowerAscii(std::string& str) {
  // str[0] is legal also if the string is empty.
  toLowerAscii(&str[0], str.size());
}
```

**Review by Yuhta — APPROVED** [review]

File: velox/functions/prestosql/types/JsonType.cpp [hunk 2]
```diff
@@ -856,30 +886,51 @@ struct CastFromJsonTypedImpl {
...
-        folly::F14FastMap<std::string, simdjson::ondemand::value> lowerCaseKeys(
-            object.count_fields());
+
+        // TODO Populate this mapping once, not per-row.
+        // Mapping from lower-case field names of the target RowType to their
+        // indices.
+        bool allFieldsAreAscii = true;
+        const auto size = rowType.size();
+        for (auto i = 0; i < size; ++i) {
+          allFieldsAreAscii &= isAscii(rowType.nameOf(i));
```
Comment (xiaoxmeng) [inline velox/functions/prestosql/types/JsonType.cpp:896]: nit: can we break out of the loop on the first non-ascii field name?

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova nice optimization. Thanks!

**Pull request closed without merging**